### ISSUE TITLE
Strange framerate issues, both with OBS-NDI and Screen Capture HX on the Output PC

### ISSUE THREAD

**Issue by rebb** [obs-ndi/obs-ndi#656]
I've been noticing that the framerate in the VODs and on the stream are nowhere near as smooth looking as they are reported to be. And i'm not sure what's going on.

This happens with both OBS-NDI providing the output on the Source PC or with Screen Capture HX.

The only difference with Screen Capture HX is, that it uses far less network bandwidth.
To compare, using the latest OBS-NDI on the Source PC uses around 13 MB/s on the network, while Screen Capture HX uses around 3 MB/s, all when using the "Maximum Bandwidth" Settings ( and even the "Ultra" Bandwidth setting when using Screen Capture HX ). All that with essentially the same settings ( 720p 50FPS plus Audio )

The game itself plays perfectly smooth, and it feels like the 50 FPS i'm running it as.

The streaming PC reports that it is streaming at 50 FPS as well, with the CPU never going near 100%, it's usually sitting at around 60% - no framedrops.

However, in the stream and in the VODs, it's obvious that this is nowhere as smooth as the 50 FPS of the game.
It even looks like it might be less than 30FPS.

What might be the reason for this ?
Is this a problem inside the latest release of OBS-NDI ?

I'm kind of out of ideas at this point on how to improve this.
I also cannot use the Sync-Mode "Source Settings" when using Screen Capture HX as the Source, because then the Audio starts to crackle for some reason, only the "Network" Sync Mode improves the Audio Problem. ( this was mentioned in #649 )

I kinda wish Newtek had a Github just so people can easily write Issues to them.

**Comment by rebb** [comment]
The game runs at 50FPS ( looks smooth ), Screen Capture HX is set to 50FPS, OBS-NDI at the streaming PC reports going at 50 FPS, even Twitch claims its 50 FPS and downloading the VOD and running ffprobe on it reports 50 FPS everywhere.

But why does it not look smooth ?
Am i hallucinating this ?

Bitrate is around 4300 for 720p@50FPS.
But shouldn't insufficient bitrate cause visual degradation instead of lower framerate ? Or am i wrong on this ?

**Comment by rebb** [comment]
Just posting this as an example. It's an ingame sequence that runs on preprogrammed camera movement. It was perfectly smooth while playing.

But in the VOD it's obviously not smooth at all.
I'm not even sure its near 30 FPS.

https://www.twitch.tv/videos/1024032956?t=0h53m7s

However when downloading these VODs, and using ffprobe on the files - it shows as 50 FPS in all the statistics.
Something is wrong here o_O

That being said i still have to do some extra tests. But right now i'm just confused.

**Comment by rebb** [comment]
Not sure if this is important, but i noticed that, when i test-recorded without audio on the gaming PC by accident ( no headphone was plugged in, so Windows had disabled the audio completely ), the video on the receiving OBS-NDI ended up much smoother.

Once audio was re-enabled, and recording the same sequence again, the strange lag came back.

**Comment by rebb** [comment]
Some more investigation results.
It seems that the Sync-Mode "Network" has good audio, but appears to introduce a framerate hit, while the other Sync-Mode "Source Settings" (?) appears to be smoother, but causes audio crackling.

**Comment by rebb** [comment]
I've had some success with experimenting with the frame-sync code from this fork : https://github.com/aquarat/obs-ndi

Using a frame-sync based approach instead of threaded capture seems to help a lot with most of these issues.
Props to https://github.com/aquarat ! :+1:

**Comment by aquarat** [comment]
hey, thanks @rebb, but that repo is very broken hahahaha. I used it for experimenting with the NDI SDK Frame Synchroniser but I also used it to investigate other things, like reading audio, video and metadata using multiple threads instead of a single thread. I suspected that there was some sort of bottleneck on my system and I thought it might be software. It turned out it was actually the machine's RAM, which was operating in single channel mode. It took me ages to discover this.

I also wanted to have ISOcording functionality - so I wanted to be able to set up one OBS instance to not activate tallies (the ISOcorder instance) and then have the tallies driven by the main OBS instance. I stuck the work on Github because I used the Azure Pipelines system to auto-build Windows builds (I built my Linux tests locally, as they're easy).

I regularly do 3 to 6 camera shoots and I was really struggling with keeping the cameras in sync in OBS, as well as frame stuttering. It became a big enough risk that I eventually stopped using OBS as the main system, but I still use OBS as a secondary mixer.

I've also found that a PCI-E, or in my case, a Thunderbolt 3 10 GbE link radically reduces CPU load for larger shoots, which can impact stuttering from NDI sources.

The NDI SDK Frame Synchroniser, as I understand it, needs NDI sources to send accurately timestamped data. My encoders, made by BirdDog, run Ubuntu Linux under the hood (not documented) and I've successfully hacked them and installed NTP daemons, which sync the clocks (it was literally ```apt install ntp```). This, combined with the NDI SDK Frame Syncer seemed to give good results, although I still had occasional stuttering. This may have been caused by the RAM bandwidth constraint, I haven't tested since the upgrade.

The NDI SDK is generally decent and well documented, but it has some bugs in it. The main one I've come across crashes OBS and seems to be caused by an encoder sending an empty video frame but marking the headers of the frame as the frame size being larger than zero (index out of bounds). The only way around this, that I could think of, was to separate out the NDI decoder into a separate process from OBS. vMix did this a short while later, they called it "Fault Encoder Isolation Mode" or FEIM.

Another problem I've come across with NDI, in both vMix and OBS: software encoders (read NDI SDK based encoders), for some reason, cause havoc in the form of severe stuttering, not just on the software encoded feed but also on other NDI feeds. As much as it irritates me I now run HDMI out from laptops into dedicated NDI encoders instead of using something like NDI Scan Converter. I've also had this issue with both OBS NDI output and vMix NDI outputs - so I believe it's specific to the NDI libs.

I've come across some other really weird things, like running two webcam-consuming applications on one machine, consuming from NDI Virtual Input will cause another instance of NDI Virtual Input on a completely different machine to have audio drop outs.

I just wanted something better than a hardware SDI mixer and I got this^. Hope this sheds some light :)

**Comment by rebb** [comment]
That sounds much more complex than my use case :D

Which is really just streaming to Twitch using a 2 PC Setup. The gaming PC uses Newtek's Screen Capture HX atm.

Without framesync i could only either get smooth video with crackling audio, or laggy video with decent audio.

**Comment by aquarat** [comment]
Haha, apologies if that comment was a bit long. The Frame Syncer does automatically resample both video and audio, which smoothes over a lot of issues :)

**Comment by rebb** [comment]
Well, even though this works pretty good with the framesync approach, it appears that for some reason, duplicate frames are arriving.

At least their timestamp is duplicated.

How would that happen though, if both sides are set to run at the same frame-rate ?
There are even duplicate frames when capturing at 50FPS while the Screen Capture HX is set to send at 60 FPS - although in that case the amount of duplicates is less than if both sides are set to 50 FPS.

Or is this just a problem that is unavoidable with NDI ?

Granted currently this is like 1 duplicate frame per second ( capturing at 50 ), with a recurring small "landslide" of duplicate frames of about 15-20 per second happening every few seconds.
These small duplicate-spikes only last 1 or 2 seconds though, and they don't duplicate the same frame 15 times.
When duplication happens, it's usually 1 or at most 2 of the previous frame.
Hope it makes sense, maybe i'm describing it too complicated.

I assume this is more of an NDI Screen Capture / Screen Capture HX related problem. Or something inherent to how NDI works.
Unfortunately for some reason an admin has to manually accept people to be able to write something on the Newtek forums ..

So i'll jot these findings down here for the time being.

**Comment by walterav1984** [comment]
@rebb 

Have you tried to stream or even record gaming output "@50fps" on NVIDIA with their builtin realtime hardware encoders and their own tools (shadow play) so without NDI or 2nd OBS setup involved? 

Is it smooth, if not can you try @60fps or even @59.94? I would'nt be surprised if the cards hardware encoders run at a different framerate than your desktop or game enige is running...

**Comment by rebb** [comment]
@walterav1984 I haven't tried that yet, would that be able to capture software-rendered things as well ?

One other thing i'm starting to wonder about using Screen Capture HX is, that it lets you choose an Encoder ( h.264 by default ).
Does that mean the image is lossy compressed/encoded, then sent over the network, only to be lossy compressed/encoded again by OBS ?

Wouldn't that degrade the image a lot ? I did notice that image quality wasn't quite where i wanted it to be for some reason.

Maybe i will ditch the whole idea of a 2 PC Streaming Setup altogether in the end.
"It seemed like a good idea at the time" - but now i'm noticing all these problems and there are no answers by people who should be "in the know" about this stuff.

**Comment by aquarat** [comment]
For a 2-PC streaming setup I wouldn't use a software encoder: I'd use either a hardware NDI encoder or a USB capture device. A good USB capture device will probably be the simplest, lowest latency and highest quality option for OBS. I only use hardware encoders/capture devices these days.

**Comment by rebb** [comment]
@aquarat : I think Screen Capture HX uses the GPU for encoding at least ( NVENC ), but there are no proper settings for bitrate or anything else. Unless you mean actual hardware ? I wanted to try and avoid getting extra hardware, but it could be an option.

**Comment by aquarat** [comment]
I have had nothing but problems with screen capture (software) solutions. It always arbitrarily bites me. A dedicated USB dongle, which has a loop out, will provide you with an uncompressed input into OBS at no performance penalty on your gaming machine. It'll be low latency and high quality. They really aren't expensive and HDMI isn't going anywhere for a while to come.

The OBS USB Capture input system is very mature/works well. Any pipeline that encodes to H264, decodes to a framebuffer and then encodes again to H264 will result in generational quality loss... but that quality loss may not be perceptible if the first stream's bitrate is higher than the second stream.

I often run multiple playback machines and for these I always use hardware capture devices, usually NDI based. I've just had so many problems with software solutions that use NDI, and this is using both OBS and competitor software.

**Comment by walterav1984** [comment]
@rebb 
> ... I'm noticing all these problems and there are **no answers** by people who should be in-the-know about this stuff...

Before getting to that sort of premature conclusion a more detailed explanation of your setup and tried strategies may attract more **in-the-know** people to this issue. Which may even conclude it might not even be a "NDI" nor "OBS" related issue. Don't get me wrong those 2 components can have a lot of issues for themselves which can be at the heart of framerate/stutter audio desync issues.

STEP1
Therefor I would first really try to verifiy your source system (game-pc) is really doing 720p@50 or even 1080p@50 on a HDMI/Displayport screen according to a Windows monitor refreshrate of 50Hz (some game engines allows to set 50fps unrelated to the windows monitor refreshrate...). Notice that Windows 10 itself can also do additional resolution/refresh convertions (check additional monitor settings)! Therefor verify the status info on your monitor/display if these signals are really what they are. If all is well and smooth at 50fps and 50Hz continue to 2nd step.

STEP2
Second are you using RGB output 8/10/12 or YUV8/10/12 on that monitor and are you using Video or Data levels 0-255 8-235 black white levels. Do you know what these mean? Do these affect brightness/contrast/colour levels (to contrasty / washed out) or performance of the "nvenc" or "uncompressed framebuffer" capture of your desktop/screen/gaming windows etc?

STEP3
Third do some recordings with NVIDIA its own game capture solution using geforce experiance/shadow play @-50fps? Use ffmpeg or VLC to determine if the framerate of the file is really 50fps. Replay / edit these files and import theme in OBS-studio all still smooth?

Four even try to use OBS on the source gaming PC, and use Game window capture (try different game engines vulkan/opengl/d3d11/12) if available and use the NVencoder inside OBS to output a stream or recording or even use OBS-NDI output a stream to the 2nd OBS machine... NVenc even has VP9 codec besides h264/h265.

> Does that mean the image is lossy compressed/encoded, then sent over the network, only to be lossy compressed/encoded again by OBS ?

Yep, since you noticed "nvenc" is at the heart of NDI screencapture HX, compared to NDI screencapture without NX both use compression and OBS decompresses sources so you can (re)composite text or other materials on it and than (re)compress that output again to a stream or recording.

About USB capture devices, most ones are cheap and compressed MJPEG/h264 and do not always allow to setup input/output framerate/resolution/colorspace etc. These days you can even get them under 20 / 10dollars. Still I would focus on the realtime NVenc usage which parameters can be set with advanced software like ffmpeg. For real uncompressed capture use HDMI/SDI capture cards like BMD intensity/decklink on a second system but as long a you have'nt verified STEP1 it may not work at all. Also notice that games run at 60FPS may not be captured on broadcast focuses captured cards so you would have stutter from 60FPS to 59.94FPS...

**Comment by walterav1984** [comment]
Do you really need 50fps, are you delivering to DVB/ASI cable / iptv / satelite-dish / terrestrial providers, what will you do if you 30/60 fps games?

**Comment by rebb** [comment]
I'm sorry if that came out wrong, but what i mean with "no answers by people in the know" is, actual people who are close to the code or know deeply how the NDI system works.

For example, a lot of threads on NDI issues on the Newtek Forums themselves go unanswered by Newtek people, but that's probably not uncommon on company forums.

Also here, there have been little to no answers on OBS-NDI Issues by the main programmers for months it seems.
Of course they might just be overworked, and i've been spoiled by some other Github issue sections.

re: Step 1
I'm not running the monitor or any other hardware at 50 Hz, this is all just about software caps. I cap the game to 50 FPS, which works as expected according to the stats recorded by MSI Afterburner / Rivatuner Statistics Server.
It feels perfectly fine and smooth while playing.

And i also cap the stream to 50 FPS. Mainly in an attempt to gain a little image quality from the same bitrate, while still having a reasonably smooth experience for the viewers.

Is it a dumb idea ? Probably.
If it ends up not working with a 2 PC NDI setup, i will just fall back to single-PC and call all this Research.
In the end, something was learned regardless.

re: Step 2
I didn't touch any settings related to color-spaces, so they are probably at the defaults of OBS and OBS-NDI ( Partial Range YUV BT.709 i think ? ). Other than that i don't recall ever having to choose anything related to color-space or color encodings.

re: Step 3
I did check the reported framerate etc. using FFMPeg or a similar tool when i had the strong framerate stutter issue from the start of this thread ( which has been largely improved by using framesync code ) - and noticed that even though it reported 50 FPS for the video, it really was not 50 FPS at all, but quite noticeably choppy and had many duplicate frames when doing frame-stepping using VLC. So i'm a little untrusting about the stats that these tools display.
Manual frame-stepping seems to be an OK way to look for duplicate frames .. maybe.

Something i did notice though, is that even though duplicate timestamps are reported by my code, there don't seem to be actual visible duplicate frames in the recorded video when frame-stepping through it - unless the video codec automatically smoothes duplicates out temporally.

For some reason the timecode field ( there is a timecode and a timestamp field ) of the frames are all 0 though, maybe the timecode would be a better indication of the actual time the current frame was created.
I saw some forum thread somewhere about this, where someone was confused why Screen Capture HX does not set the timecodes, while Screen Capture ( non HX ) does.

This is why i would prefer some input by people who deeply know the code ( ie developers of Screen Capture HX ), as they could comment on whether the timestamp data is meaningful or should be taken with two grains of salt. Such input could save a lot of confusion and guesswork imo.

It's very possible that i just totally misunderstand or misinterpret what's actually going on.

re: Step 4
This is kind of my last resort option and the one i might go for after all. Simply streaming directly from the same PC.

I don't see any NVEnc Options for other Codecs anywhere in OBS though. Screen Capture HX offers "HEVC" as an option, but that does not work for some reason.
Maybe that would be something to investigate further when it comes to trying to combat the image-quality degradation of 2 encodings taking place.

Do i need 50 FPS ?
It's what i want to target for streaming at least.
It's probably dumb, but so am i :smiley:

**Comment by walterav1984** [comment]
If you cannot verify STEP1, than I would just try for testing purposes the otherway around to run everything (whole pipeline from game source/monitor to OBS project frequency and streaming framerate) at 59,94Hz or maybe 60Hz native (without artificial capping) if that resembles your monitor and see how that works.

Remember NDI comes from Newtek a broadcast focused company, although things might work they work better and are tested at 50/59.94 fps maybe nondrop frames to for ntsc at 60fps. NDI is designed to handle loses to some extend, so it will perform some magic just to keep continuity.

Perceptually you are right that 50fps might be a good way to prefer compression efficiency compared to higher 59.94 or 60fps framerates with lower quality frames. This has been and is still used in PAL countries around the world to show smooth motion with a little higher resolution in SD times. But remember if a codec has good temporal compression efficiency besides spatial you might not even notice the difference per frame when going from 59.94 > 50 fps at the same bitrate.

**Comment by mutant-industries** [comment]
The same issue is with my setup and honestly it really sucks. I'm using OBS Camera app on iPhone 12 connected to router via GBit 802.11ax. I'm trying to receive NDI output (720p60) on linux PC (OBS 27.2, libndi.so.4.6.2, obs-ndi v4.9.0) and Win10 PC (OBS 29.0.2,  NDI 5.5.3 Runtime, obs-ndi v4.11.0) and both have the same issue as described above:
- when 'source timing' is selected in NDI source I get constant distortion and glitches in source audio, but the video has no issues
- when 'network timing' is selected in NDI source the sound is flawless, but the video loses frames from time to time. Sometimes every two following frames are duplicated, sometimes it runs just fine, but usually for no longer that for one minute. 

This problem exists definitely on receiving side (obs-ndi), because when I receive the same NDI stream via NDI Studio Monitor and try to record it then it is just flawless.

I remember using the linux machine (as described) with the OBS Camera app sometime during Feb - May 2022, I recorded and streamed many videos using the 'source timing' and looking back at those videos I don't see any audio or video issues. Therefore there must have been some update in NDI library within OBS Camera app since then.

Anyway here is my workaround:
- I create one NDI source just for audio - in bandwidth I select audio only, Sync Network with latency mode low (experimental)
- I create second NDI source just for video - I disable NDIPlugin.SourceProps.Audio, Sync Source Timing and again latency mode low (experimental)
- In audio mixer I mute and hide audio track, that belongs to video source (it is silent anyway)
- In advanced audio properties I set sync offset 20ms to audio source

Somehow this setup seems to work without issues even during longer streams (I do usually 2 hours) on both computers.
